src/dag.c

```diff
@@ -304,19 +531,40 @@ void RedisAI_DagRunSessionStep(RedisAI_RunInfo *rinfo, const char *devicestr,
     }
   }
 
-  pthread_mutex_lock(rinfo->dagMutex);
-
-  if (currentOp->result == REDISMODULE_OK) {
-    *progress = 1;
+  if (currentOp->result != REDISMODULE_OK) {
+    __atomic_store_n(rinfo->dagError, 1, __ATOMIC_RELAXED);
   }
-  else {
-    *progress = 0;
-    *rinfo->dagError = 1;
+  
+  return;
+}
+
+void RedisAI_BatchedDagRunSessionStep(RedisAI_RunInfo **batched_rinfo, const char *devicestr) {
+  // Assumption: ops are guaranteed to be all MODELRUN
+
+  int n_ops = array_len(batched_rinfo);
+
+  assert(n_ops > 1);
+
+  RAI_DagOp *currentOps[n_ops];
+
+  for (int i=0; i<n_ops; i++) {
+    RedisAI_RunInfo *rinfo = batched_rinfo[i];
+
+    RAI_DagOp *currentOp = RedisAI_DagCurrentOp(rinfo);
+
+    currentOps[i] = currentOp;
   }
 
-  *rinfo->dagRefCount -= 1;
+  RedisAI_BatchedDagRunSession_ModelRun_Step(batched_rinfo, currentOps);
+
+  for (int i=0; i<n_ops; i++) {
+    RedisAI_RunInfo *rinfo = batched_rinfo[i];
+    RAI_DagOp *currentOp = currentOps[i];
 
-  pthread_mutex_unlock(rinfo->dagMutex);
+    if (currentOp->result != REDISMODULE_OK) {
+      __atomic_store_n(rinfo->dagError, 1, __ATOMIC_RELAXED);
```

Comment:
Here we are copying the error to all ops participating in the batch, so that they carry the info that an error occurred while they were being executed. I think this is cleaner rather than having to keep track of this at the worker level.